### ISSUE TITLE
parametric type on only some of variants

### ISSUE THREAD

**Issue by disberd** [Roger-luo/Moshi.jl#11]
Hello,

I was trying out the package and wanted to reproduce the very simple example in the README of DynamicSumTypes.jl to compare synthax/performance.

That example (slightly adapted) creates a sumtype like so:
```julia
using DynamicSumTypes
abstract type AbstractAT end
struct A{X}
    x::X
end
struct B
    y::Float64
end
@sumtype AT(A{Int}, B) <: AbstractAT
```

I tried creating something similar with Moshi doing this:
```julia
@data AT{X} begin
	struct A
		x::X
	end
	struct B
		y::Float64
	end
end
```

The problem with the Moshi version is that `AT.B` has no constructor, it seems a constructor for `B` is correctly generated only in the following two cases:
- `B` has no fields
- `B` has at least one field of type `X`

Is there any alternative way to create variants that do not use the parametric type but still have fields?

**Comment by disberd** [comment]
Thanks for the very detailed explanation, it is clearer now.

I think it would be worth to add some example/explanation about this in the docs, what do you think?

**Comment by Roger-luo** [comment]
Definitely! As you can see, the documentation still needs to be completed. (that's why I created the gotcha label, in case I forget)
